# Mobile Apps Analysis

Guided Project: Profitable App Profiles for the App Store and Google Play Markets

The goal of the project is to analyse data to determine which apps users would prefer to use.

Mobile apps are: 
* available for English users
* free to download and install. 


[Apple apps store dataset](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)

[Google play store apps dataset](https://www.kaggle.com/datasets/lava18/google-play-store-apps)

In [1]:
from csv import reader

apple_file = open("AppleStore.csv", encoding="utf8")
read_apple_file = list(reader(apple_file))

google_file = open("googleplaystore.csv", encoding="utf8")
read_google_file = list(reader(google_file))

In [2]:
"""
A function to explore dataset

Parameters
    dataset : list
        The data to be read
    start : int
        Index of the first row
    end : int
        Index of the last row

Returns
    Nothing
"""
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
#explore_data(read_apple_file, 1, len(read_apple_file))

In [4]:
#explore_data(read_apple_file, 1, len(read_apple_file))

In [5]:
# Data with header
print(len(read_apple_file))
print(len(read_google_file))

7198
10842


In [6]:
google_header = read_google_file[0]
google_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [7]:
apple_header = read_apple_file[0]
apple_header

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [8]:
# Remove header
apple_data = read_apple_file[1:]
google_data = read_google_file[1:]

In [9]:
print(f"Length of apple data: {len(apple_data)}")
print(f"Length of google data: {len(google_data)}")

Length of apple data: 7197
Length of google data: 10841


# Data  Cleaning

Let's start by removing row with missing entry from the google playstore dataset.

In [10]:
# check for row with missing column entry
for row in google_data:
    if len(row) != len(google_data[0]):
        print(google_data.index(row), len(row))

10472 12


In [11]:
google_data[10472]

['Life Made WI-Fi Touchscreen Photo Frame',
 '1.9',
 '19',
 '3.0M',
 '1,000+',
 'Free',
 '0',
 'Everyone',
 '',
 'February 11, 2018',
 '1.0.19',
 '4.0 and up']

In [12]:
read_google_file[0]

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [13]:
del google_data[10472] # delete row with error

In [14]:
len(google_data)

10840

In [15]:
# check for duplicates
unique_rows = []
duplicate_rows = []

for row in google_data:
    name = row[0]
    
    if name in unique_rows:
        duplicate_rows.append(name)
    unique_rows.append(name)
print(len(duplicate_rows))
print(len(unique_rows))


1181
10840


In [16]:
print(duplicate_rows[0:5])

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


In [17]:
print(f"Expected length: ",len(google_data) - len(duplicate_rows))

Expected length:  9659


In [18]:
read_google_file[0]

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [19]:
# Remove duplicates from the dataset

reviews_max = {} # dictionary to hold apps with highest reviews

for row in google_data: # google_data is the google play store dataset without header
    name = row[0]
    n_reviews = float(row[3]) # float needed when comparing reviews to get the highest 
    
    # check if app exist in the dictionary and the previous review in the dictionary is lesser than the new review
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews    
    # otherwise create a review for it in the dictionary
    if name not in reviews_max:
        reviews_max[name] = n_reviews
# use the dictionary of max reviews to clean the google play dataset
android_clean = [] # to store new cleaned dataset
already_added = [] # to store app names

for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)


Note: android_clean is the cleaned goole playstore data. 

Check if apple ios store data has duplicates.

In [20]:
print(apple_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [21]:
duplicate_apple_rows = []
unique_apple_rows = []

for row in apple_data:
    id = row[0]
    if id in duplicate_apple_rows:
        duplicate_apple_rows.append(id)
    unique_apple_rows.append(id)

In [22]:
print("SIZES OF APPLE DATA")
print("-------------------")
print(f"apple data: {len(apple_data)}")
print(f"duplicate data: {len(duplicate_apple_rows)}")
print(f"unique data: {len(unique_apple_rows)}")

SIZES OF APPLE DATA
-------------------
apple data: 7197
duplicate data: 0
unique data: 7197


From the look of things, the apple data does not have duplicates.

## Non-English apps

Some of the apps contain information that are not in English. We are trying to analyse data on mobile apps for English users.

In [23]:
print(apple_data[813][1])
print(android_clean[4412][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
中国語 AQリスニング


Let's remove the apps with names that contain non-English symbols. The ASCII range of English characters are 0-127.

In [24]:
# Use ord() to check ASCII equivalent of a string character
print(ord("A"))
print(ord("+"))

65
43


Since ord() takes in only characters, loop through a string to read the ASCII equivalent of the entire string. For instance:

In [25]:
for letter in "dog":
    print(ord(letter))

100
111
103


In [26]:
"""
A function to remove rows corresponding to non-English apps

Parameters
    a_string : string
        The input string

Return
    False : Boolean
        The output of the function if condition is met

"""
def check_english(a_string):
    for element in a_string:
        if ord(element) < 0 or ord(element) > 127:
            return False  

In [27]:
check_list = ['Instagram',
'爱奇艺PPS -《欢乐颂2》电视剧热播',
'Docs To Go™ Free Office Suite',
'Instachat 😜']

for element in check_list:
    print(check_english(element))

None
False
False
False


In [28]:
print(ord("😜"))

128540


The check_english() function could not detect apps with English names that contain emojis. 

*A not too perfect hack:*

Let's edit the check_english() function to only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range.

In [29]:
"""
A function to check if rows corresponding to non-English apps exist

Parameters
    a_string : string
        The input string

Return
    False : Boolean
        The output of the function if condition is met

"""
def english_checker(a_string):
    count = 0
    for element in a_string:
        if ord(element) < 0 or ord(element) > 127:
            count +=1
        if count > 3:
            return False  

In [30]:
check_language = ['Instagram',
'爱奇艺PPS -《欢乐颂2》电视剧热播',
'Docs To Go™ Free Office Suite',
'Instachat 😜']

for element in check_list:
    print(english_checker(element))

None
False
None
None


Remove apps from **apple_data** and **android_clean** datasets with non-English names using the english_checker function. 

In [31]:
print(apple_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [32]:
print(google_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [33]:
# Apple
ios_data = []
bad_ios_data = []
for row in apple_data:
    track_name = row[1]
    
    if english_checker(track_name) == False:
        bad_ios_data.append(row)
    if english_checker(track_name) == None:
        ios_data.append(row)

In [34]:
len(apple_data)

7197

In [35]:
len(ios_data)

6183

In [36]:
len(bad_ios_data)

1014

In [37]:
# Google
android_data = []
bad_android_data = []
for row in android_clean:
    app_name = row[0]
    
    if english_checker(app_name) == False:
        bad_android_data.append(row)
    if english_checker(app_name) == None:
        android_data.append(row)

In [38]:
len(android_clean)

9659

In [39]:
len(android_data)

9614

In [40]:
len(bad_android_data)

45

After checking for English or non-English apps:

Size of android_data: 9614

Size of ios_data: 6183

## Remove non-free apps

In [41]:
google_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [42]:
free_android = []

for row in android_data:
    if row[7] == "$0.0" or row[7] == "0.0" or row[7] == "0":
        free_android.append(row)

In [43]:
print(len(free_android))

8864


In [44]:
apple_header

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [45]:
free_ios = []
for row in ios_data:
    if row[4] == "$0.0" or row[4] == "0.0" or row[4] == "0":
        free_ios.append(row)       

In [46]:
print(len(free_ios))

3222


Size of free_android: 8864

Size of free_ios: 3222

## Data analysis

The goal of the project is to build mobile apps for both google store and iOS store. e are trying to analyse apps on these stores to see what attributes they possess that attracts users.


After cleaning the data, we have two datasets:

* free_android, length 8864 (header: google_header)

* free_ios, length 3222 (header: apple_header)

To generate frequency table (in percent), the **prime_genre** (at index 11) column is useful for the free_ios data, while the **Category** (at index 1) and **Genres** (at index 9) columns of the free_android data.

The frequency table (a dictionary) can be constructed using the algorithm:

* create an empty dictionary

* initialise a total as zero

* loop through the data
    
    * increment total for every row to get the total entries
    
    * increment row value if found in the dictionary
    
    * if not found, create a row value by assigning 1 to that value.
    

The percentages table can use the created frequency table (dictionary) to get percentage values for all key-value pairs using  value of key / total * 100

In [47]:
# Create frequency table
# Google play

def frequency_table(dataset, index): # index is the column number in the data
    freq_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]

        if value in freq_table:
            freq_table[value] +=1
        else:
            freq_table[value] = 1
    
    percentage_table = {}
    
    for key in freq_table:
        percent = freq_table[key] / total * 100
        
        percentage_table[key] = percent
        
    return percentage_table

In [48]:
frequency_table(free_ios, 11) # prime_genre

{'Social Networking': 3.2898820608317814,
 'Photo & Video': 4.9658597144630665,
 'Games': 58.16263190564867,
 'Music': 2.0484171322160147,
 'Reference': 0.5586592178770949,
 'Health & Fitness': 2.0173805090006205,
 'Weather': 0.8690254500310366,
 'Utilities': 2.5139664804469275,
 'Travel': 1.2414649286157666,
 'Shopping': 2.60707635009311,
 'News': 1.3345747982619491,
 'Navigation': 0.186219739292365,
 'Lifestyle': 1.5828677839851024,
 'Entertainment': 7.883302296710118,
 'Food & Drink': 0.8069522036002483,
 'Sports': 2.1415270018621975,
 'Book': 0.4345127250155183,
 'Finance': 1.1173184357541899,
 'Education': 3.662321539416512,
 'Productivity': 1.7380509000620732,
 'Business': 0.5276225946617008,
 'Catalogs': 0.12414649286157665,
 'Medical': 0.186219739292365}

In [49]:
android_freq = frequency_table(free_android, 9)

In [50]:
android_freq

{'Art & Design': 0.5979241877256317,
 'Art & Design;Creativity': 0.06768953068592057,
 'Auto & Vehicles': 0.9250902527075812,
 'Beauty': 0.5979241877256317,
 'Books & Reference': 2.1435018050541514,
 'Business': 4.591606498194946,
 'Comics': 0.6092057761732852,
 'Comics;Creativity': 0.01128158844765343,
 'Communication': 3.2378158844765346,
 'Dating': 1.861462093862816,
 'Education': 5.347472924187725,
 'Education;Creativity': 0.04512635379061372,
 'Education;Education': 0.33844765342960287,
 'Education;Pretend Play': 0.056407942238267145,
 'Education;Brain Games': 0.033844765342960284,
 'Entertainment': 6.069494584837545,
 'Entertainment;Brain Games': 0.078971119133574,
 'Entertainment;Creativity': 0.033844765342960284,
 'Entertainment;Music & Video': 0.16922382671480143,
 'Events': 0.7107400722021661,
 'Finance': 3.7003610108303246,
 'Food & Drink': 1.2409747292418771,
 'Health & Fitness': 3.0798736462093865,
 'House & Home': 0.8235559566787004,
 'Libraries & Demo': 0.936371841155234

In [51]:
fin_max = max(android_freq, key=android_freq.get)
print("Maximum value:",fin_max)

Maximum value: Tools


In [52]:
def display_table(dataset, index):
    table = frequency_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [53]:
display_table(free_ios, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


In [54]:
display_table(free_android, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [55]:
display_table(free_android, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

To determine most popular genres, let's begin by calculating average number of user ratings (on the App Store)

In [67]:
freq_prime_genre = frequency_table(free_ios, 11)

In [68]:
freq_prime_genre

{'Social Networking': 3.2898820608317814,
 'Photo & Video': 4.9658597144630665,
 'Games': 58.16263190564867,
 'Music': 2.0484171322160147,
 'Reference': 0.5586592178770949,
 'Health & Fitness': 2.0173805090006205,
 'Weather': 0.8690254500310366,
 'Utilities': 2.5139664804469275,
 'Travel': 1.2414649286157666,
 'Shopping': 2.60707635009311,
 'News': 1.3345747982619491,
 'Navigation': 0.186219739292365,
 'Lifestyle': 1.5828677839851024,
 'Entertainment': 7.883302296710118,
 'Food & Drink': 0.8069522036002483,
 'Sports': 2.1415270018621975,
 'Book': 0.4345127250155183,
 'Finance': 1.1173184357541899,
 'Education': 3.662321539416512,
 'Productivity': 1.7380509000620732,
 'Business': 0.5276225946617008,
 'Catalogs': 0.12414649286157665,
 'Medical': 0.186219739292365}

In [71]:
free_ios[11]

['359917414',
 'Solitaire',
 '101943296',
 'USD',
 '0.0',
 '679055',
 '9673',
 '4.5',
 '4.5',
 '4.11.2',
 '4+',
 'Games',
 '38',
 '4',
 '11',
 '1']

In [77]:
apple_header

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [78]:
free_ios[0]

['284882215',
 'Facebook',
 '389879808',
 'USD',
 '0.0',
 '2974676',
 '212',
 '3.5',
 '3.5',
 '95.0',
 '4+',
 'Social Networking',
 '37',
 '1',
 '29',
 '1']

In [86]:
freq_prime_genre = frequency_table(free_ios, 11)
for genre in freq_prime_genre:
    total = 0 # sum of the number of ratings (not sum of ratings)
    len_genre = 0 # store num of apps wrt each genre

    for row in free_ios:
        genre_app = row[11]

        if genre_app == genre:
            num_ratings = float(row[5])
            total += num_ratings
            len_genre += 1
            
    avg_user_ratings = total / len_genre
    print(f"{genre} : {avg_user_ratings}")

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Let's compute number of installs per app genre for Google Play

In [87]:
google_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [93]:
cat_freq = frequency_table(free_android, 1)
for category in cat_freq:
    total = 0
    len_category = 0
    
    for row in free_android:
        category_app = row[1]
        
        if category_app == category:
            num_installs = row[5]
            num_installs = num_installs.replace("+", "")
            num_installs = num_installs.replace(",", "")
            num_installs = float(num_installs)
            total += num_installs
            len_category += 1
    avg_nun_installs = total / len_category 
    print(f"{category} : {avg_nun_installs}")

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_